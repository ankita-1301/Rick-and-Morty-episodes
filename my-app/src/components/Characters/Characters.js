import React from "react";
import "../../styles.css";
import "antd/dist/antd.css";
import { Card, Avatar, Button } from "antd";
import { UserOutlined } from "@ant-design/icons";
import { Link } from "react-router-dom";
import { getCharacter } from "rickmortyapi";

//Component for rendering all the characters
export default class Characters extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      characters: [],
      next: 0, //counter if nextPage exists for all characters
    };
  }

  componentDidMount() {
    this.fetchAllCharacters();
  }

  //get all the characters and update its next page with a counter
  fetchAllCharacters = () => {
    getCharacter()
      .then(({ info, results }) => {
        this.setState({
          characters: results,
          next: !!info.next ? this.state.next + 1 : this.state.next,
        });
      })
      .catch((err) => console.log(err));
  };

  //load more characters when click on button and update next counter
  onClickLoadMore = () => {
    getCharacter({ page: this.state.next + 1 })
      .then(({ info, results }) => {
        this.setState({ characters: [...this.state.characters, ...results] });
        this.setState({ next: !!info.next ? this.state.next + 1 : 0 });
      })
      .catch((err) => console.log(err));
  };

  render() {
    const { characters, next } = this.state;

    const gridStyle = {
      width: "25%",
      textAlign: "center",
      padding: "20px",
    };

    //card-grid for all the characters with its link to CharacterView
    let cardData = characters.map((data) => {
      return (
        <Link
          to={{
            pathname: `/characters/character/${data.id}`,
            state: { data: data }, //character details
          }}
          key={data.id}
        >
          <Card.Grid style={gridStyle} key={data.id}>
            <Avatar
              shape="square"
              size={100}
              icon={<UserOutlined />}
              src={data.image}
            />
            <h1>{data.name}</h1>
          </Card.Grid>
        </Link>
      );
    });

    return (
      <div>
        <Card
          title="All characters of Ricky and Morty episodes"
          data-test="characters-card"
        >
          {cardData}
          {!!next && (
            <Button className="load-more-button" onClick={this.onClickLoadMore}>
              Load more...
            </Button>
          )}
        </Card>
      </div>
    );
  }
}
