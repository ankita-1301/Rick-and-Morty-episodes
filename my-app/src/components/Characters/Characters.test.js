import React from "react";
import Enzyme, { shallow } from "enzyme";
import Characters from "./Characters";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";

Enzyme.configure({ adapter: new Adapter() });

test("Characters component snapshot", () => {
  const characters = renderer.create(<Characters />);
  let charTree = characters.toJSON();
  expect(charTree).toMatchSnapshot();
});

const findByAttri = (component, attri) => {
  const wrapper = component.find(`[data-test="${attri}"]`);
  return wrapper;
};

const setUp = (props = {}) => {
  const component = shallow(<Characters {...props} />);
  return component;
};

describe("Test characters component", () => {
  describe("Has prop title", () => {
    let wrapper;
    beforeEach(() => {
      const props = {
        title: "All characters of Ricky and Morty episodes",
      };
      wrapper = setUp(props);
    });

    it("Renders characters without error", () => {
      const component = findByAttri(wrapper, `characters-card`);
      expect(component.length).toBe(1);
    });
  });
});
