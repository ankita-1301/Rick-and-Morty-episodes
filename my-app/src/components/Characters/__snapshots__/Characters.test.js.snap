// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters component snapshot 1`] = `
<div>
  <div
    className="ant-card ant-card-bordered"
    data-test="characters-card"
  >
    <div
      className="ant-card-head"
      style={Object {}}
    >
      <div
        className="ant-card-head-wrapper"
      >
        <div
          className="ant-card-head-title"
        >
          All characters of Ricky and Morty episodes
        </div>
      </div>
    </div>
    <div
      className="ant-card-body"
      style={Object {}}
    />
  </div>
</div>
`;
